### ISSUE TITLE
Declaration emit redundantly inlines optional union parameters

### ISSUE THREAD

**Issue by robpalme** [microsoft/TypeScript#42165]
# Bug Report

### 🔎 Search Terms

inlining duplication declaration emit optional union parameters

### 🕗 Version & Regression Information

- This is not a regression.
- This is a stylistic issue in emitted DTS.  The code is correct.
- This is the behavior in every version I tried (3.3 ... nightly)

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBArlC8UCMUA+UBMaoGZsBZsBWAbigFgAoKiADzAHsAnYKAYwYDsBnVhhKAG8qUUVAC2AQ04ATScGYgAFJKYBzAFywAlFoBuDAJYyhqtQF8ANKID0NqABEAKgGUoTCADMIHzmwjcsJyGXFBKagwMMtoiYgxgwCGckgA2KuoA-FowulAGxqbqVrb2zm6GnCkVAUFJYQBGktFU5lRAA)

### 💻 Code

```ts
type u = 1 | 2 | 3 | 4 | 5; 

export const o = {
    mandatory(arg: u): void {arg},  // DTS references union (good)
    optional(arg?: u): void {arg},  // DTS inlines union (bad)
}
```

### 🙁 Actual behavior

DTS emit expands the definition of the union for the optional parameter.

```ts
declare type u = 1 | 2 | 3 | 4 | 5;
export declare const o: {
    mandatory(arg: u): void;
    optional(arg?: 1 | 2 | 3 | 4 | 5 | undefined): void;
};
```

### 🙂 Expected behavior

DTS emit should treat it the same as a mandatory parameter and reference it by name.

```ts
declare type u = 1 | 2 | 3 | 4 | 5;
export declare const o: {
    mandatory(arg: u): void;
    optional(arg?: u): void;
};
```

### Related Issues

This is a very specific instance of [the issue of declarations inlining types](https://github.com/microsoft/TypeScript/issues/37151).  I'm raising it because it looks like a pure win to simplify emitted DTS files.  Fixing this will make other userland tricks to simplify types more effective.

If fixing this is contentious, or requires a large overhaul, please feel free to immediately close.  I will not be offend because I realize there is a long tail of issues in this space and do not wish to create noise.

**Comment by andrewbranch** [comment]
I bet @weswigham will know off the top of his head how feasible this is.

**Comment by weswigham** [comment]
Specifically for optional property serialization, it should be feasible to see if the `undefined`-stripped union as an alias symbol, and to use that if available.

**Comment by OliverJAsh** [comment]
Related (it's not just unions which include `undefined`): https://github.com/microsoft/TypeScript/issues/26106

**Comment by weswigham** [comment]
Yeah, the `undefined` case is just very common and easy to scan for. To generally fix it, we'd have to generate all possible groups of sub-unions for a union (including those with duplicates shared between sets), and see if any combinations of subsets have "better" printback (where "better" likely means shorter), and then choose the best of the available options - and even that wouldn't capture the aliases originally in unions which subtype reduced when combined (as those subtypes, and the unions they were originally in, are likely irrecoverable). So the general case isn't really easy to solve for.

**Comment by OliverJAsh** [comment]
Will both of these issues be fixed by https://github.com/microsoft/TypeScript/pull/42149?

**Comment by robpalme** [comment]
Closing because the original code example is now fixed in TypeScript 4.2 by https://github.com/microsoft/TypeScript/pull/42149

There's still a redundant `| undefined` in the emit, but I think that can be considered a different issue.

Thanks for acting on this @andrewbranch (even if you did get Sherlocked by Anders)
